cip/1.accepted/CIP2017-02-06-Path-Patterns.adoc

```diff
@@ -0,0 +1,439 @@
+
+[source, cypher]
+.A pattern matching a path of length 0
+----
+MATCH (a)-[()]-(b)
+----
+
+This pattern states that `a` and `b` must be the same node, by virtue of stating a pattern that matches any node.
+It is thus the same as:
+
+[source, cypher]
+----
+MATCH (a), (b) WHERE a = b
+----
+
+The Path Patterns start becoming interesting when larger expressions are put together:
+
+[source, cypher]
+.Finding someone loved by someone hated by someone you know, transitively
+----
+MATCH (you)-[[:KNOWS :HATES]+ :LOVES]->(someone)
+----
+
+Note the `+` expressing one or more occurrences of the sequence `KNOWS` followed by `HATES`.
+
+The direction of each relationship is governed by the overall direction of the Path Pattern.
+It is however possible to explicitly define the direction for a particular part of the pattern.
+This is done by either prefixing that part with `<` for a right-to-left direction or suffix it with `>` for a left-to-right direction.
+It is possible to both prefix the part with `<` and suffix it with `>`, indicating that this part of the pattern matches in any direction.
+
+[source, cypher]
+.Specifying the direction for different parts of the pattern
+----
+MATCH (you)-[[:KNOWS <:HATES]+ :LOVES]->(someone)
+----
+
+In the example above we say that the `HATES` relationships should have the opposite direction to the other relationships in the path.
+
+Through the use of Defined Path Predicates we can express even more predicates over a path:
+
+[source, cypher]
+.Find a chain of unreciprocated lovers
+----
+PATH PATTERN unreciprocated_love = (a)-[:LOVES]->(b)
+     WHERE NOT EXISTS { (b)-[:LOVES]->(a) }
+MATCH (you)-[~unreciprocated_love*]->(someone)
+----
+
+Note how there is no colon used for referencing the Defined Path Predicate, the colon is used in Path Patterns only for referencing actual relationship types.
+
+Sometimes it will be interesting to express a predicate on a node in a Path Pattern.
+This can be achieved by using a Defined Path Predicate where the nodes on both ends are the same:
+
+[source, cypher]
+.Find friends of friends that are not haters
+----
+PATH PATTERN not_a_hater = (x)
+     WHERE NOT EXISTS { (x)-[:HATES]->() }
+MATCH (you)-[:KNOWS ~not_a_hater :KNOWS]-(friend_of_friendly_friend)
+----
+
+In the case of a Defined Path Predicate where both nodes are the same, the direction of the predicate is irrelevant.
+In general the direction of a Defined Path Predicate is quite important, and used for mapping the pattern in the predicate into the Path Patterns that reference it.
+The only cases where it is allowed to omit the direction of a Defined Path Predicate is when the defined predicate is reflexive.
+This is obviously the case when both nodes are the same, but it would also be the case when the internal pattern is symmetrical, such as in the following example:
+
+[source, cypher]
+.Find chains of co-authorship
+----
+PATH PATTERN co_author = (a)-[:AUTHORED]->(:Book)<-[:AUTHORED]-(b)
```

Comment:
But it is an example of when the named path pattern _itself_ is undirected - and the direction *is* left out on the next row.